Utilities/ocvalidate/ocvalidate.c

```diff
@@ -34,41 +37,1272 @@
-    // printf("milliseconds: %lld\n", milliseconds);
-    return milliseconds;
+typedef
+UINT32
+(*CONFIG_CHECK) (
+  IN  OC_GLOBAL_CONFIG  *Config
+  );
+
+STATIC
+INT64
+GetCurrentTimestamp (
+  VOID
+  )
+{
+  struct timeval Time;
+  //
+  // Get current time.
+  //
+  gettimeofday (&Time, NULL);
+  //
+  // Return milliseconds.
+  //
+  return Time.tv_sec * 1000LL + Time.tv_usec / 1000LL;
 }
 
-int main(int argc, char** argv) {
-  uint32_t f;
-  uint8_t *b;
-  if ((b = readFile(argc > 1 ? argv[1] : "config.plist", &f)) == NULL) {
-    printf("Read fail\n");
-    return -1;
+STATIC
+CHAR8 *
+GetFilenameSuffix (
+  IN  CONST CHAR8  *FileName
+  )
+{
+  CONST CHAR8  *SuffixDot;
+
+  //
+  // Find the last dot by reverse order.
+  //
+  SuffixDot = OcAsciiStrrChr (FileName, '.');
+
+  //
+  // In some weird cases the filename can be crazily like
+  // "a.suffix." (note that '.') or "..." (extremely abnormal).
+  // We do not support such.
+  //
+  if (!SuffixDot || SuffixDot == FileName) {
+    return (CHAR8 *) "";
+  }
+
+  //
+  // In most normal cases, return whatever following the dot.
+  //
+  return (CHAR8 *) SuffixDot + 1;
+}
+
+STATIC
+BOOLEAN
+AsciiStringHasAllLegalCharacter (
+  IN  CONST CHAR8  *String
+  )
+{
+  UINTN  Index;
+
+  for (Index = 0; Index < AsciiStrLen (String); ++Index) {
+    //
+    // Skip allowed characters (0-9, A-Z, a-z, '_', '-', '.', '/', and '\').
+    //
+    if (IsAsciiNumber (String[Index])
+      || IsAsciiAlpha (String[Index])
+      || String[Index] == '_'
+      || String[Index] == '-'
+      || String[Index] == '.'
+      || String[Index] == '/'
+      || String[Index] == '\\') {
+      continue;
+    }
+
+    //
+    // Disallowed characters matched.
+    //
+    return FALSE;
   }
 
-  long long a = current_timestamp();
+  return TRUE;
+}
+
+STATIC
+BOOLEAN
+AsciiStringHasAllPrintableCharacter (
+  IN  CONST CHAR8  *String
+  )
+{
+  UINTN  Index;
+
+  for (Index = 0; Index < AsciiStrLen (String); ++Index) {
+    //
+    // Unprintable characters matched.
+    //
+    if (IsAsciiPrint (String[Index]) == 0) {
+      return FALSE;
+    }
+  }
+
+  return TRUE;
+}
+
+STATIC
+UINT32
+ReportError (
+  IN  CONST CHAR8  *FuncName,
+  IN  UINT32       ErrorCount
+  )
+{
+  if (ErrorCount != 0) {
+    DEBUG ((DEBUG_WARN, "%a returns %u %a!\n", FuncName, ErrorCount, ErrorCount > 1 ? "errors" : "error"));
+  } else {
+    DEBUG ((DEBUG_VERBOSE, "%a returns no errors!\n", FuncName));
+  }
+
+  return ErrorCount;
+}
+
+STATIC
+BOOLEAN
+DataHasProperMasking (
+  IN  CONST VOID   *Data,
+  IN  CONST VOID   *Mask,
+  IN        UINTN  Size
+  )
+{
+  CONST UINT8  *ByteData;
+  CONST UINT8  *ByteMask;
+  UINTN        Index;
+
+  ByteData = Data;
+  ByteMask = Mask;
+
+  for (Index = 0; Index < Size; ++Index) {
+    if ((ByteData[Index] & ~ByteMask[Index]) != 0) {
+      return FALSE;
+    }
+  }
+
+  return TRUE;
+}
+
+STATIC
+UINT32
+CheckACPI (
```

Comment:
The file gets large, please separate checkers to separate files. For example, you could make ocvalidate.h header, and move checkers to separate files: ValidateAcpi.c, ValidateBooter, etc.

Reply:
Done. https://github.com/PMheart/OpenCorePkg/commit/56b5cab4d9f7ed30373580f8b7e6b45078e34ff0